File: priv/javascript/lib/map_set.js
import SpecialForms from './kernel/special_forms';
import List from './list';

function __new__(){
  return SpecialForms.map({ [Symbol.for("__struct__")]: Symbol.for("MapSet"), set: SpecialForms.list() });
}

function size(map){
  return keys(map.set).length;
}

function to_list(map){
  return map.set;
}

function __delete__(set, term){
  let new_list = List.__delete__(set.set, term);

  let new_map = Object.assign({}, set);
  new_map.set = new_list;
  return SpecialForms.map(new_map); 
}

function put(set, term){
  if(set.set.indexOf(term) === -1){
    let new_list = List.append(set.set, term);

    let new_map = Object.assign({}, set);
    new_map.set = new_list;
    return SpecialForms.map(new_map); 
  }

  return set;
}

function difference(set1, set2){
  let new_map = Object.assign({}, set1);

  for(let val of set1.set){
    if(member__qmark__(set2, val)){
      new_map.set = List.__delete__(new_map.set, val);
    }
  }

  return SpecialForms.map(new_map); 
}


function intersection(set1, set2){
  let new_map = Object.assign({}, set1);

  for(let val of set1.set){
    if(!member__qmark__(set2, val)){
      new_map.set = List.__delete__(new_map.set, val);
    }
  }

  return SpecialForms.map(new_map); 
}


function union(set1, set2){
  let new_map = set1;

  for(let val of set2.set){
      new_map = put(set, val);
  }

  return SpecialForms.map(new_map); 
}


function disjoin__qmark__(set1, set2){
  for(let val of set1.set){
    if(member__qmark__(set2, val)){
      return false;
    }
  }

  return true;
}

function member__qmark__(set, value){
  return set.set.indexof(value) >= 0;
}

function equal__qmark__(set1, set2){
  return set1.set === set2.set;
}

function subset__qmark__(set1, set2){
  for(let val of set1.set){
    if(!member__qmark__(set2, val)){
      return false;
    }
  }

  return true;
}


export default {
  __new__,
  size,
  to_list,
  disjoin__qmark__,
  __delete__,
  subset__qmark__,
  equal__qmark__,
  member__qmark__,
  put,
  union,
  intersection,
  difference
}

File: lib/elixir_script/prelude/map_set.ex
defmodule ElixirScript.MapSet do

  defstruct set: []

  def new() do
    %MapSet{}
  end

  def size(set) do
    length(set.set)
  end

  def to_list(set) do
    set.set
  end

  def delete(set, term) do
    %{ set | set: Elixir.List.remove(set.set, term) }
  end

  def put(set, term) do
    case member?(set, term) do
      false ->
        %{ set | set: set.set ++ term }
      true ->
        set
    end
  end

  def member?(set, term) do
    set.set.indexOf(term) >= 0
  end

  def equal?(set1, set2) do
    set1 === set2
  end

  def difference(set1, set2) do
    do_difference(to_list(set1), set2, new())
  end

  def do_difference([], _, difference_set) do
    difference_set
  end

  def do_difference(set1_list, set2, difference_set) do
    term = hd(set1_list)
    case member?(set2, term) do
      true ->
        do_difference(tl(set1_list), set2, difference_set)
      false ->
        do_difference(tl(set1_list), set2, %{ difference_set | set: difference_set.set ++ [term]})
    end
  end

  def intersection(set1, set2) do
    do_intersection(to_list(set1), set2, new())
  end

  def do_intersection([], _, intersection_set) do
    intersection_set
  end

  def do_intersection(set1_list, set2, intersection_set) do
    term = hd(set1_list)
    case member?(set2, term) do
      false ->
        do_intersection(tl(set1_list), set2, intersection_set)
      true ->
        do_intersection(tl(set1_list), set2, %{ intersection_set | set: intersection_set.set ++ [term]})
    end
  end

  def union(set1, set2) do
    %{ set1 | set: set1.set ++ set2.set}
  end

  def disjoint?(set1, set2) do
    size(intersection(set1, set2)) == 0
  end

  def subset?(set1, set2) do
    do_subset?(to_list(set1), set2)
  end

  def do_subset?([], _) do
    true
  end

  def do_subset?(set1_list, set2) do
    term = hd(set1_list)
    case member?(set2, term) do
      false ->
        false
      true ->
        do_subset?(tl(set1_list), set2)
    end
  end

end
